### ISSUE TITLE
Why is it that when using prulong to eval, the code in duoattention is adopted instead？

### ISSUE THREAD

**Issue by qqtang-code** [princeton-pli/PruLong#6]
Great Work!!
I noticed that when evaluating PruLong, the code appears to invoke components from DuoAttention. Specifically, it seems that llama_duo_attention_forward_one_way_reordered is being used during the prefill phase, which employs full attention rather than the streaming head. As I understand, in the original DuoAttention design, the prefill stage uses full attention for initial context processing, while the streaming head is reserved for the decoding/generation phase.
```python
    if q_len == kv_seq_len:
        # pre-filling: use flash attention
        attn_output = flash_attn_func(
            query_states,
            key_states,
            value_states,
            causal=True,
            dropout_p=0.0,
        )
    else:
        # decoding or continous filling
        if self.num_full_attn_head > 0:
            full_query_states = query_states[:, :, : self.num_full_query_head, :]

            full_attn_output = flash_attn_func(
                full_query_states,
                full_key_states,
                full_value_states,
                causal=True,
                dropout_p=0.0,
            )
        else:
            full_attn_output = None

        if self.num_streaming_attn_head > 0:
            streaming_query_states = query_states[:, :, self.num_full_query_head :, :]

            streaming_attn_output = flash_attn_func(
                streaming_query_states,
                streaming_key_states,
                streaming_value_states,
                causal=True,
                dropout_p=0.0,
            )
        else:
            streaming_attn_output = None

        if full_attn_output is None:
            attn_output = streaming_attn_output
        elif streaming_attn_output is None:
            attn_output = full_attn_output
        else:
            attn_output = torch.cat([full_attn_output, streaming_attn_output], dim=2)
```

Although PruLong does show significant improvements over DuoAttention in terms of performance. However, it still lags noticeably behind full attention baselines. Even compared to other sparse attention methods like XAttention, the performance gap remains substantial. While PruLong achieves a much smaller memory footprint and computational cost—making it promising for long-sequence applications—the accuracy degradation is non-negligible and could limit its practical utility in quality-sensitive scenarios.

**Comment by testzer0** [comment]
Hi qqtang,
Thanks for the query! Since the output of PruLong is a set of masks for the model, we can use the DuoAttention framework for performing inference. During prefilling, we have to materialize the KVs for the entire current chunk anyway, so using local attention would not yield any gains. Therefore, we allow full attention within the chunk (and the KV footprint is correspondingly higher to reflect this). 

As for full attention baselines: I am not sure where you are getting the XAttention comparison from? The XAttention paper does not evaluate on the tasks we do, and as we see from the results in our [paper](https://arxiv.org/abs/2506.17121), unless we evaluate across a set of diverse tasks (including long-form generation) it is hard to get an accurate picture of which method is best. You may also refer to Section 2 of our paper for a discussion of why sparse full attention methods may not lead to lower KV footprints, which is what we target here.

Feel free to re-open if you have other questions,
Adithya
